new/cmd/k8s-nsmgr/pkg/deviceplugin/server.go

```diff
@@ -0,0 +1,154 @@
+package deviceplugin
+
+import (
+	"context"
+	"net"
+	"net/url"
+	"strconv"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/crossapi/chains/nsmgr"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/crossapi/chains/nsmgr/peer_tracker"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/tools/serialize"
+	"github.com/networkservicemesh/networkservicemesh/pkg/tools"
+	"github.com/networkservicemesh/networkservicemesh/sdk/common"
+	"google.golang.org/grpc"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+)
+
+type NsmDevicePluginServer interface {
+	networkservice.NetworkServiceServer
+	pluginapi.DevicePluginServer
+}
+
+type nsmgrDevicePlugin struct {
+	nsmgr.Nsmgr
+	devices               map[string]*pluginapi.Device
+	allocatedDevices      map[string]*pluginapi.Device
+	executor              serialize.Executor
+	insecure              bool
+	reallocate            func(u *url.URL)
+	listAndWatchListeners []pluginapi.DevicePlugin_ListAndWatchServer
+}
+
+func NewServer(name string, insecure bool, registryCC *grpc.ClientConn) NsmDevicePluginServer {
+	rv := &nsmgrDevicePlugin{
+		devices:          make(map[string]*pluginapi.Device, DeviceBuffer),
+		allocatedDevices: make(map[string]*pluginapi.Device, DeviceBuffer),
+		executor:         serialize.NewExecutor(),
+		insecure:         insecure,
+	}
+	// TODO - Fix applying peer_tracker here
+	// rv.Nsmgr = peer_tracker.NewServer(nsmgr.NewEndpoint(name, registryCC), &rv.reallocate)
+	rv.Nsmgr = peer_tracker.NewServer(nsmgr.NewNsmgr(name, registryCC), &rv.reallocate)
+	rv.resizeDevicePool()
+	return rv
+}
+
+func (n *nsmgrDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	n.executor.Exec(func() {
+		n.listAndWatchListeners = append(n.listAndWatchListeners, s)
+		listAndWatchResponse := &pluginapi.ListAndWatchResponse{}
+		for _, device := range n.devices {
+			listAndWatchResponse.Devices = append(listAndWatchResponse.Devices, device)
+		}
+		for _, listAndWatchListener := range n.listAndWatchListeners {
+			listAndWatchListener.Send(listAndWatchResponse)
+		}
+	})
+	select {
+	case <-s.Context().Done():
```

Comment:
Looks like we can get rid of using select here because we have only one channel here.

Reply:
Excellent point!  Fixed :)